### ISSUE TITLE
[coq makefile] Plugin linking strategy is broken and will doubly-link modules, rendering many plugins broken in OCaml >= 4.08.0

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#12067]
As can be witnessed in #11580 coq makefile will doubly link modules in `.cmxs` files which is forbidden in OCaml >= 4.08.0

This breaks some plugins, such as mtac2.

**Comment by SkySkimmer** [comment]
Isn't it coqc doing the linking, not coq_makefile?

**Comment by ejgallego** [comment]
> Isn't it coqc doing the linking, not coq_makefile?

By linking I mean building the `.cmxs` archive; Coq just call dynlink so it is too late for it to do something.

**Comment by Janno** [comment]
Any idea what plugin developers can do about this? I am getting reports by people who are unable to install Unicoq on 4.09.

**Comment by Janno** [comment]
It seems this https://github.com/smtcoq/smtcoq/pull/69/files#diff-4a9c012d0765dd7f558a5ef81b3ba40dR43 is a backwards-compatible workaround. It feels rather silly, though. I couldn't manage to overwrite CAMLDONTLINK from _CoqProject so it seems I actually need a `configure` script as they have in smtcoq?

**Comment by ejgallego** [comment]
I'm much afraid that until someone has the time to tackle the underlying Coq linker issue https://github.com/coq/coq/issues/7698 there is little to do. Maybe vendoring unicoq using dune and actually making it a private library would work.

**Comment by ejgallego** [comment]
@Janno that will only work for num as it is already linked by coqtop. That's another solution, you could link unicoq statically.

In your case actually if you manage to have coq_makefile not link the plugin [or dune], then you can load unicoq manually and that should work. For one plugin is still acceptable [but this doesn't scale for more complex plugins such as elpi]

**Comment by gares** [comment]
In master the coq_makefile variable `CAMLDONTLINK=str,unix,dynlink,threads,zarith` seems up to date. These modules, already linked in the top level, are not also linked in plugins.

Is there still a problem? Which is it exactly? CC @beta-ziliani you say you have a recent log with the problem at hand?

**Comment by ejgallego** [comment]
The problem is for any plugin depending on a plugin, for example `mtac2` will link unicoq twice; as outline coq_makefile should not use -linkall in cmxs otherwise any plugin that uses an overlapping lib will be broken. mtac2 could workaround this by actually not loading unicoq, but that's a bit of a gross hack.

One strategy to solve this is to use dynlink support from findlib, François Bobot implemented support for this in Dune, however coq_makefile would need a few changes to generate METAs for plugins.

**Comment by gares** [comment]
I think I see the problem in abstract. But locally, I can build/run Mtac2 just fine with with OCaml 4.10.0, and I can do it in two ways.

In its standard setup Mtac2 does not link Unicoq via findlib (there is no `-package Unicoq`, the Makefile.local adds a `-I` by hand in order to find unicoq's cmi, via the not exactly orthodox `COQ_SRC_SUBDIRS+=user-contrib/Unicoq
`). Anyway, it compiler fine on 8.12.0 on 4.10.0 and I don't see in the logs an attempt to link unicoq.cm* in MetaCoqPlugin.cmxs.

I tried to remove the line changing COQ_SRC_SUBDIRS, and instead add
```
CAMLPKGS+= -package Unicoq
CAMLDONTLINK+=,Unicoq
export OCAMLPATH=/home/gares/.opam/4.10.0/lib/coq/user-contrib/
```
(and I did place a META in /home/gares/.opam/4.10.0/lib/coq/user-contrib/Unicoq/META). Things still work fine.

So I'm not so sure I'm able to reproduce the problem. Do I need 4.11? (a message up there talk of 4.08+)

**Comment by gares** [comment]
I think I can make the linking bug happen, on purpose, but it does not seem it affects Mtac2. @beta-ziliani could you post the log?

**Comment by beta-ziliani** [comment]
I have another issue compiling Unicoq, I don't know if it's related:
```
Error: Dynlink error: interface mismatch on Vmvalues
```

**Comment by beta-ziliani** [comment]
I can compile it with no problem as is (by hand). The issue I raised is not in the compilation of Mtac2, but of coq.dev from opam. I'm using Ocaml 4.09.1.

**Comment by beta-ziliani** [comment]
the problem is in the CI, so I need to check there what version of ocaml it's using. will come back with that info.

**Comment by Zimmi48** [comment]
We have faced this issue multiple times in the past. For instance, try to run the Windows packaging job with add-ons in master and you will get the error.

**Comment by ejgallego** [comment]
@gares is this fixed by #15220 too?

**Comment by gares** [comment]
yes, camldontlink is gone.

**Comment by ejgallego** [comment]
Closed by #15220
